django/db/models/query_utils.py

```diff
@@ -175,6 +175,26 @@ def __eq__(self, other):
     def __hash__(self):
         return hash(self.identity)
 
+    @cached_property
+    def referenced_base_fields(self):
+        """
+        Retrieve all base fields referenced directly or through F expressions excluding
+        any fields referenced through joins.
+        """
+        # Avoid circular imports.
+        from django.db.models import F
+
+        expressions = list(self.flatten())
+        f_references = {expr.name for expr in expressions if isinstance(expr, F)}
+        q_references = {
+            child[0].split(LOOKUP_SEP, 1)[0]
+            for expr in expressions
+            if isinstance(expr, Q)
+            for child in expr.children
+            if isinstance(child, tuple)
+        }
+        return f_references | q_references
```

Comment:
Can we reuse `django.db.models.sql.query.get_children_from_q()`? The only difference is that it doesn't stop on the first lookup. If we change `q.referenced_base_fields` to the `set(get_children_from(q))` only two cases fail with: `field_7__lookup` instead of `field_7`.

Reply:
Yes, looks like we can 👍 this would reduce to
```python
    @cached_property
    def referenced_base_fields(self):
        """
        Retrieve all base fields referenced directly or through F expressions
        excluding any fields referenced through joins.
        """
        # Avoid circular imports.
        from django.db.models.sql import query

        return set(
            map(lambda x: x.split(LOOKUP_SEP, 1)[0], query.get_children_from_q(self))
        )
```